odoo/tools/translate.py

```diff
@@ -179,6 +182,15 @@ def translate_attrib_value(node):
 avoid_pattern = re.compile(r"\s*<!DOCTYPE", re.IGNORECASE | re.MULTILINE | re.UNICODE)
 
 
+def should_force_translatable_element(node):
+    # Some specific nodes (e.g., text highlights) have an auto-updated DOM
+    # structure that makes them impossible to translate. The introduction of a
+    # translation `<span>` in the middle of their hierarchy breaks their
+    # functionalities. The goal here is to force them to be translated as a
+    # whole using their CSS class selectors.
+    classes = node.attrib.get("class", "").split()
+    return any(translated_class in classes for translated_class in TRANSLATED_ELEMENTS_CLASSES)
```

Comment:
@HydrionBurst What do you think of doing something like this?

IMO there is a more fondamental problem surrounding the translation of a sentence which contains links: if you have a sentence like this:
`word_1 link_1 word_2 link_2 word_3` this translatable as 5 different parts instead of a sentence because of this very old change: https://github.com/odoo/odoo/commit/9bd60ca93510e410a0136b8b433f596330900593 ... should not this be changed?

Reply:
The idea is interesting and should be helpful in many business use cases
Technically I prefer to use a more technical class/attribute name instead of the class "o_text_highlight"
[+] "o_text_highlight" is more like a business name for style instead of a technical name for translation
[+] translators are less likely to remove it while translating
[+] easier to know everything is translated as a whole when debug
[+] everything can be done in base module. (If website is uninstalled, the change of TRANSLATED_ELEMENTS_CLASSES may cauase html struature inconsitency)
[+] also works for `word_1 link_1 word_2 link_2 word_3` problem.
[-] new class/attribute name doesn't work for already created html
[-] inthe edit_translations mode, it may still have things like `'<span><div class="o_text_highlight">text<div></span>'`. Not sure if it acceptable
Another approach is to add some exceptions to the <a>
e.g. `<a inline-translation="1" href="`

cc @rco-odoo

Reply:
@HydrionBurst 

> I prefer to use a more technical class/attribute
> [...]
> [-] new class/attribute name doesn't work for already created html

Yes I agree on the principle but we really need to fix our customer databases. It's rather violent: without this, any highlight with a link inside just prevents translating anything on the page (not just the highlight), and the users have no way to know why :grimacing: Because the editor is really not happy of having span pop up in those weird places :grimacing: 

> Another approach is to add some exceptions to the `<a>` e.g. `<a inline-translation="1" href="`

Not sure this is the best approach in an editor where you could create the `<a>` then maybe later wrap it into a highlight, then later remove the highlight etc etc.

> [+] also works for word_1 link_1 word_2 link_2 word_3 problem.

What works? The only way to make that work is to re-add the `<a>` element as a `TRANSLATED_ELEMENTS` IMO... and find another way for the original usecase that was fixed by removing the `<a>` from that list (like maybe, just prevent the exact usecase of ul > li > a that was discussed in the original commit)... although this is likely something that we cannot change in stable :grimacing:

Reply:
After 16.0, once an element is added to TRANSLATED_ELEMENTS_CLASSES
or like this commit, a class is added to TRANSLATED_ELEMENTS_CLASSES
you should never remove it. Otherwise, the translation mapping might be ruined
e.g. 
```py
{
	'en_US': '<div class="o_text_highlight">text1 text2</div>',
        # [+] also works for word_1 link_1 word_2 link_2 word_3 problem.
	'fr_FR': '<div class="o_text_highlight">text1<a href="link">text2</a></div>'
	
}
```
when "o_text_highlight" is in TRANSLATED_ELEMENTS_CLASSES
it has en_US term `'<div class="o_text_highlight">text1 text2</div>'`
and fr_FR term `'<div class="o_text_highlight">text1<a href="link">text2</a></div>'`

but after "o_text_highlight" is removed from TRANSLATED_ELEMENTS_CLASSES
en_US term becomes: 'text1 text2'
fr_FR terms are: 'text1' and 'text2'

Maybe we use o_text_highlight for TRANSLATED_ELEMENTS_CLASSES in stable
And use another tag e.g. class="o_translation_inline" with a migration script for elements with class="o_text_highlight" in master

Reply:
> Not sure this is the best approach in an editor where you could create the <a> then maybe later wrap it into a highlight, then later remove the highlight etc etc.

@rco-odoo do you remember if https://github.com/odoo/odoo/commit/9bd60ca93510e410a0136b8b433f596330900593 blocks `<a/>` only for menu? Is it ok to allow other or some `<a/>`s to be translated inline?

Reply:
From what I remember, translating the `<a>` elements inline was specifically breaking some cases, and we reacted by discarding the inline translation of those elements altogether.  The current PR simply shows that some `<a>` elements should legitimately be translated inline, but fixes the issue by adding an extra layer on complexity.  Maybe we should simply allow the inline translation of `<a>` elements **except for the ones** that are known to be problematic.

Reply:
@rco-odoo cc @HydrionBurst That's what I meant by

> and find another way for the original usecase that was fixed by removing the `<a>` from that list

Yes :pray: 

Should we handle that?

Reply:
I don't know what is the side effect of the `<span>` wrapped `<a>` for the menu in the context `edit_translations`. But from the translate function's point of view, it is safe to add `<a>` back without any upgrade script.
Maybe your team can handle that?
If we do it for 17.0, maybe we also have to reexport pot files and retranslate some terms cc @oomsveta

Reply:
I am not sure that would safe enough to do in 17.0 though... :grimacing: Would not it mean losing translations at best? Breaking translating cases at worse? (original point of the fix that removed `<a>` from the list, maybe more).

Should not we go ahead with this PR and reflect on this in master?

Reply:
If we decide to `allow the inline translation of `<a>` elements except for the ones that are known to be problematic`, we can only do it in master since we need a special tag for those `<a>` exceptions which needs to change the views